src/bidiMapper/modules/context/NavigationTracker.ts

```diff
@@ -19,50 +19,69 @@
 import type {Protocol} from 'devtools-protocol';
 
 import {
+  type BrowsingContext,
   ChromiumBidi,
-  UnknownErrorException,
 } from '../../../protocol/protocol.js';
 import {Deferred} from '../../../utils/Deferred.js';
+import {type LoggerFn, LogType} from '../../../utils/log.js';
 import {getTimestamp} from '../../../utils/time.js';
 import {urlMatchesAboutBlank} from '../../../utils/UrlHelpers.js';
 import {uuidv4} from '../../../utils/uuid.js';
 import type {EventManager} from '../session/EventManager.js';
 
+export type NavigationEventName =
+  | ChromiumBidi.BrowsingContext.EventNames.FragmentNavigated
+  | ChromiumBidi.BrowsingContext.EventNames.NavigationAborted
+  | ChromiumBidi.BrowsingContext.EventNames.NavigationFailed
+  | ChromiumBidi.BrowsingContext.EventNames.Load;
+
 class NavigationState {
   readonly navigationId = uuidv4();
-  url?: string;
+  readonly #browsingContextId: string;
+
+  started = new Deferred<void>();
+  finished = new Deferred<NavigationEventName>();
+  url: string;
+  loaderId?: string;
 
-  constructor(url?: string) {
+  constructor(url: string, browsingContextId: string) {
+    this.#browsingContextId = browsingContextId;
     this.url = url;
+    void this.finished.then(() => {
+      this.started.reject(
+        new Error('Navigation finished without being started'),
+      );
+      return;
+    });
+  }
+
+  navigationInfo(): BrowsingContext.NavigationInfo {
+    return {
+      context: this.#browsingContextId,
+      navigation: this.navigationId,
+      timestamp: getTimestamp(),
+      url: this.url,
+    };
   }
 }
 
+/**
+ * Keeps track of navigations. Details: http://go/webdriver:bidi-navigation
+ */
 export class NavigationTracker {
```

Comment:
could we cover major use cases with unit tests?

Reply:
added the major cases